src/lightning/fabric/fabric.py

```diff
@@ -764,9 +768,9 @@ def _run_impl(self, run_method: Callable, *args: Any, **kwargs: Any) -> Any:
     def _run_with_setup(self, run_function: Callable, *args: Any, **kwargs: Any) -> Any:
         self._strategy.setup_environment()
         # apply sharded context to prevent OOM
-        with self.sharded_model(), _replace_dunder_methods(DataLoader, "dataset"), _replace_dunder_methods(
-            BatchSampler
-        ):
+        with _old_sharded_model_context(self._strategy), _replace_dunder_methods(
```

Comment:
So in theory applying it twice shouldn't hurt. If you want, I can write a test for it to make sure nothing funky happens with FSDP/DeepSpeed.

Reply:
Ideally, we should remove the _old_sharded_model_context in favor of control, because the user can't opt out of it. This is an artifact of the LightningLite.run() design.

Reply:
I agree on your last comment. Since both deepspeed and fsdp are experimental, we are free to do this removal

Reply:
There wouldn't be any impact for current Fabric end-users (not developers) right? If so, I agree we should remove.

Reply:
There wouldn't be an impact, because this is only used for the old LightningLite.run() construct which is not documented for Fabric and only used in like 3 of our tests, unless there is a LightningLite user who could have migrated to Fabric without removing their code from the Fabric.run() method and it would still work.

Reply:
It's not just those using `.run()` but also `.launch(fn)` as it also wraps with `_run_with_setup`

Reply:
Yes, which is why I kept it and the discussion for removal can be done separately in another PR, probably easier after #17479